onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.cc

```diff
@@ -338,35 +345,43 @@ Status DQMatMulToMatMulNBitsAction::ProcessNewNode(Graph& graph,
   // to what we need. But it does not handle external data.
   Initializer weight_src(*weight_tensor_proto, graph.ModelPath());
   Initializer scale_src(*scale_tensor_proto, graph.ModelPath());
-  std::optional<Initializer> zp_src;
-  Initializer weight_dst(ONNX_NAMESPACE::TensorProto_DataType_UINT8,
-                         graph.GenerateNodeArgName(weight_arg->Name() + "_T"),
-                         std::vector<int64_t>{N, quant_num, blob_bytes});
-  Initializer scale_dst(static_cast<ONNX_NAMESPACE::TensorProto_DataType>(scale_src.data_type()),
-                        graph.GenerateNodeArgName(scale_arg->Name() + "_T"),
-                        std::vector<int64_t>{N * quant_num});
-  std::optional<Initializer> zp_dst;
+  auto uint8_type = DataTypeImpl::TensorTypeFromONNXEnum(ONNX_NAMESPACE::TensorProto_DataType_UINT8)->GetElementType();
+  auto scale_type = DataTypeImpl::TensorTypeFromONNXEnum(scale_src.data_type())->GetElementType();
+  std::optional<Initializer> zp_src_ptr;
+  auto weight_dst_name = graph.GenerateNodeArgName(weight_arg->Name() + "_T");
+  auto weight_dst_ptr = std::make_unique<Tensor>(uint8_type,
+                                                 TensorShape{N, quant_num, blob_bytes},
+                                                 std::make_shared<CPUAllocator>());
+  auto scale_dst_name = graph.GenerateNodeArgName(scale_arg->Name() + "_T");
+  auto scale_dst_ptr = std::make_unique<Tensor>(scale_type,
+                                                TensorShape{N * quant_num},
```

Comment:
N * quant_num could overflow. You should use SafeInt, or let TensorShape class handle it.

Reply:
update in next iter